File: src/Accipere/RAM.hs
module Accipere.RAM where

-- /proc/meminfo
-- 1 - MemTotal:       32789044 kB
-- 2 - MemFree:        16709236 kB
-- 3 - MemAvailable:   27133040 kB
-- 15 - SwapTotal:      16394236 kB
-- 16 - SwapFree:       16394236 kB

data RAM = RAM
  { total     :: Int,
    free      :: Int,
    available :: Int,
    swapTotal :: Int,
    swapFree  :: Int
  }

instance Show RAM where
  show (RAM {..}) =
    "[RAM]:"
      <> "\n\tTotal: "
      <> show total
      <> "\n\tFree: "
      <> show free
      <> "\n\tAvailable: "
      <> show available
      <> "\n\tSwap: "
      <> "\n\t\tTotal: "
      <> show swapTotal
      <> "\n\t\tFree: "
      <> show swapFree

getRAMInfo :: IO RAM
getRAMInfo = do
  meminfo <- readFile "/proc/meminfo"
  let cleanedUp = map go $ lines meminfo
  return $
    RAM
      { total = read (cleanedUp !! 0) :: Int,
        free = read (cleanedUp !! 1) :: Int,
        available = read (cleanedUp !! 2) :: Int,
        swapTotal = read (cleanedUp !! 14) :: Int,
        swapFree = read (cleanedUp !! 15) :: Int
      }
  where
    go :: String -> String
    go [] = []
    go line =
      case words line of
        (_ : value : _) -> value
        []              -> ""
        _               -> ""

File: src/ram.rs
use std::{
    fs,
};

#[derive(Debug)]
pub struct Ram {
    pub total: u64,
    pub available: u64,
}

impl Ram {
    pub fn new() -> Self {
        let mut total = 0;
        let mut available = 0;

        let mut props: u8 = 0;
        if let Ok(meminfo) = fs::read_to_string("/proc/meminfo") {
            for line in meminfo.lines() {
                if props >= 2 {
                    break;
                };
                match line {
                    line if line.starts_with("MemTotal:") => {
                        props += 1;
                        total = line
                            .split_whitespace()
                            .nth(1)
                            .unwrap_or_default()
                            .parse::<u64>()
                            .unwrap_or_default()
                    }
                    line if line.starts_with("MemAvailable:") => {
                        props += 1;
                        available = line
                            .split_whitespace()
                            .nth(1)
                            .unwrap()
                            .parse::<u64>()
                            .unwrap()
                    }
                    _ => continue,
                }
            }
        };
        Ram { total, available }
    }
}
